consensus/polybft/consensus_runtime.go

```diff
@@ -0,0 +1,851 @@
+			}
+
+			if err := c.state.cleanCommitments(nextStateSyncExecutionIdx); err != nil {
+				return nil, err
+			}
+
+			nonExecutedCommitments, err := c.state.getNonExecutedCommitments(nextStateSyncExecutionIdx)
+			if err != nil {
+				return nil, err
+			}
+
+			if len(nonExecutedCommitments) > 0 {
+				bundlesToExecute, err := c.state.getBundles(nextStateSyncExecutionIdx, maxBundlesPerSprint)
+				if err != nil {
+					return nil, err
+				}
+
+				ff.commitmentsToVerifyBundles = nonExecutedCommitments
+				ff.bundleProofs = bundlesToExecute
+			}
+		}
+	}
+
+	// TODO: Nemanja
+	// if isEndOfEpoch {
+	// 	uptimeCounter, err := c.calculateUptime(parent)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+
+	// 	ff.uptimeCounter = uptimeCounter
+	// }
+
+	c.logger.Info("[FSM built]",
+		"epoch", epoch.Number,
+		"endOfEpoch", isEndOfEpoch,
+		"endOfSprint", isEndOfSprint,
+	)
+
+	return ff, nil
+}
+
+// restartEpoch resets the previously run epoch and moves to the next one
+func (c *consensusRuntime) restartEpoch(header *types.Header) error {
+	c.lastBuiltBlock = header
+
+	systemState, err := c.getSystemState(c.lastBuiltBlock)
+	if err != nil {
+		return err
+	}
+
+	epochNumber, err := systemState.GetEpoch()
+	if err != nil {
+		return err
+	}
+
+	lastEpoch := c.getEpoch()
+	if lastEpoch != nil {
+		// Epoch might be already in memory, if its the same number do nothing.
+		// Otherwise, reset the epoch metadata and restart the async services
+		if lastEpoch.Number == epochNumber {
+			return nil
+		}
+	}
+
+	/*
+		// We will uncomment this once we have the clear PoC for the checkpoint
+		lastCheckpoint := uint64(0)
+
+		// get the blocks that should be signed for this checkpoint period
+		blocks := []*types.Block{}
+
+		epochSize := c.config.Config.PolyBFT.Epoch
+		for i := lastCheckpoint * epochSize; i < epoch*epochSize; i++ {
+			block := c.config.Blockchain.GetBlockByNumber(i)
+			if block == nil {
+				panic("block not found")
+			} else {
+				blocks = append(blocks, block)
+			}
+		}
+	*/
+
+	validatorSet, err := c.config.polybftBackend.GetValidators(c.lastBuiltBlock.Number, nil)
+	if err != nil {
+		return err
+	}
+
+	epoch := &epochMetadata{
+		Number:         epochNumber,
+		LastCheckpoint: 0,
+		Blocks:         []*types.Block{},
+		Validators:     validatorSet,
+	}
+
+	if err := c.state.cleanEpochsFromDB(); err != nil {
+		c.logger.Error("Could not clean previous epochs from db.", "err", err)
+	}
+
+	if err := c.state.insertEpoch(epoch.Number); err != nil {
+		return fmt.Errorf("an error occurred while inserting new epoch in db. Reason: %w", err)
+	}
+
+	// create commitment for state sync events
+	if c.IsBridgeEnabled() {
+		nextCommittedIndex, err := systemState.GetNextCommittedIndex()
+		if err != nil {
+			return err
+		}
+
+		commitment, err := c.buildCommitment(epochNumber, nextCommittedIndex)
+		if err != nil {
+			return err
+		}
+
+		epoch.Commitment = commitment
+	}
+
+	c.lock.Lock()
+	c.epoch = epoch
+	c.lock.Unlock()
```

Comment:
You've already defined a helper for fetching, please define it for setting as well